Preserve group leader in Logger.ErrorHandler notification to Logger

wrachwal opened the pull request:
Logger.ErrorHandler which basically forwards received event (changing its shape, but this is irrelevant) should always preserve original group leader (gl). It's Logger's handlers (backends) which can do whatever they want with gl, based on their own strategies.

Similarly, suppressing events based on node difference (node(gl) != node()) is too early, IMO. The event had to travel to the event manager (:error_loger), so one hop to the final manager (Logger) is almost at no cost, as both reside in the same  node. So let the backends to decide.

josevalim commented:
> Similarly, suppressing events based on node difference (node(gl) != node()) is too early, IMO. The event had to travel to the event manager (:error_loger), so one hop to the final manager (Logger) is almost at no cost, as both reside in the same node. So let the backends to decide.

The default error_logger does the node redirecting, so we could either do that too, or leave it up to the Logger, as you mentioned. So basically we have three options:

1. discard it
2. redirect to the other node
3. leave it up to the Logger backend

I would like to hear more about the pros and cons of each and which flexibilities we would win in case we leave it up to the Logger.

wrachwal commented:
@josevalim good points. Erlangers had to have good reasons to redirect. Thinking otherwise would be unfair ;) Let's take simplest case - playing with a process spawned in a shell connected to a remote node. With 3) `:error_logger` events would be suppressed completely, only 2) will assure to work with console backend as it is today. Volume of such stream is typically low, but not seeing it at all - very annoying (it happened to me once to stumble across on similar case).

josevalim commented:
So I think we should redirect as well, exactly as Erlang's error_logger.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

wrachwal commented:
Hmm, is that at all necessary? If `:error_logger` (as handler) is still on the list, wouldn't it do what we wanted (and always did??)?
```elixir
iex(1)> :gen_event.which_handlers :error_logger
[Logger.ErrorHandler, :error_logger]
```
```erlang
handle_event({Type, GL, Msg}, State) when node(GL) =/= node() ->
    gen_event:notify({error_logger, node(GL)},{Type, GL, Msg}),
    %% handle_event2({Type, GL, Msg}, State);  %% Shall we do something
    {ok, State};                               %% at this node too ???
```

josevalim commented:
Oh I thought we removed that handler but we only remove the tty one. You are right. Doesn't it then mean that the message is arriving to the local node anyway? So shouldn't we also discard it to avoid logging it twice?

wrachwal commented:
Yes, exactly. I reverted the clause with event discard - and rewrote the commit... don't know whether it's good or not. github didn't punish me, but I don't see any button to re-test. Updated commit is `be00c50`.

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
